ffn/utils/proofreading.py

```diff
@@ -182,58 +260,293 @@ def __init__(self, objects, bad, num_to_prefetch=10, locations=None):
+      self.seg_error_coordinates.update({identifier: self.temp_coord_list})
+      self.annotate_error_locations(self.temp_coord_list, identifier)
+      self.temp_coord_list = []
+
+  def annotate_error_locations(
+      self, coordinate_lst: list[list[int]], id_: str
+  ) -> None:
+    """Annotate the error locations in the viewer.
+
+    Args:
+      coordinate_lst: List of coordinates to be annotated.
+      id_: Unique identifier for the error.
+    """
+    for i, coord in enumerate(coordinate_lst):
+      annotation_id = id_ + f"_{i}"
+      self.mk_point_annotation(coord, annotation_id)
+
+  def mk_point_annotation(
+      self, coordinate: list[int], annotation_id: str
+  ) -> None:
+    """Create a point annotation in the viewer.
+
+    Args:
+      coordinate: 3D coordinate of the annotation point.
+      annotation_id: Unique identifier for the annotation.
+    """
+    if annotation_id.startswith("m"):
+      color = "#fae505"
+    else:
+      color = "#05f2fa"
+    annotation = neuroglancer.PointAnnotation(
+        id=annotation_id, point=coordinate, props=[color]
+    )
+    with self.viewer.txn() as s:
+      annotations = s.layers["annotation"].annotations
+      annotations.append(annotation)
+
+  def get_annotation_id(
+      self, action_state: neuroglancer.viewer_config_state.ActionState
+  ) -> Optional[str]:
+    """Retrieve the ID of a selected annotation.
+
+    Args:
+      action_state: neuroglancer.viewer_config_state.ActionState.
+
+    Returns:
+      The selected object's ID or None if retrieval fails.
+    """
+    try:
+      selection_state = action_state.selected_values["annotation"].to_json()
+      selected_object = selection_state["annotationId"]
+    except Exception:
+      self.update_msg("Could not retrieve annotation id")
+      return
+
+    return selected_object
+
+  def delete_location_from_annotation(
+      self, action_state: neuroglancer.viewer_config_state.ActionState
+  ) -> None:
+    """Delete the error location pair associated with the annotation at the cursor position
+
+    Args:
+      action_state: State of the viewer during the action.
+    """
+    id_ = self.get_annotation_id(action_state)
+    if id_ is None:
+      return
+
+    target_key = id_[:2]
+    del self.seg_error_coordinates[target_key]
+
+    to_remove = [target_key + "_0", target_key + "_1"]
+    self.delete_annotation(to_remove)
+
+  def delete_annotation(self, to_remove: list[str]) -> None:
+    """Delete specified annotations from the viewer.
+
+    Args:
+      to_remove: list of annotation IDs to be removed.
+    """
+    with self.viewer.txn() as s:
+      annotations = s.layers["annotation"].annotations
+      annotations = [a for a in annotations if a.id not in to_remove]
+      s.layers["annotation"].annotations = annotations
+
+  def delete_last_location(self):
+    """Delete the last error location pair tagged."""
+    last_key = list(self.seg_error_coordinates.keys())[-1]
```

Comment:
Prefer: next(reversed(self.seg_error_coordinates)) for faster lookup of the last key.